File: day21/missing_number.py
import random

def missing_number(nums):
    n = len(nums)
    expected_sum = n * (n + 1) // 2  # sum of 0 to n
    return expected_sum - sum(nums)

def generate_test_case():
    n = random.randint(10, 30)
    full = list(range(n + 1))  # 0 to n inclusive
    missing = random.choice(full)
    full.remove(missing)
    random.shuffle(full)
    return full, missing

def main():
    nums, actual_missing = generate_test_case()
    print("Input:", nums)
    print("Expected missing number:", actual_missing)
    print("Computed missing number:", missing_number(nums))

if __name__ == "__main__":
    main()

File: day21/missing_number.rs
use rand::{rng, random_range, seq::SliceRandom};

fn missing_number(nums: &[usize]) -> usize {
    let n = nums.len();
    let expected_sum = n * (n + 1) / 2;
    let actual_sum: usize = nums.iter().sum();
    expected_sum - actual_sum
}

fn main() {
    let mut rng = rng();
    let n = random_range(10..=30);

    let mut full: Vec<usize> = (0..=n).collect();
    let missing_index = random_range(0..=n);
    let missing = full.remove(missing_index);
    full.shuffle(&mut rng);

    println!("Input: {:?}", full);
    println!("Expected missing number: {}", missing);
    println!("Computed missing number: {}", missing_number(&full));
}
